### PR TITLE
Na lms kat multi level

### PR DESCRIPTION
**Pull request by ashman-p**
This PR contains 2-Level LMS support. This allows for large numbers of signatures (millions).
Added 2-level variants that are supported by WolfSSL and BC.
Included KAT from RFC 8554 (LMS).
  

* [ ] Does this PR change the input/output behaviour of a cryptographic algorithm (i.e., does it change known answer test values)?  (If so, a version bump will be required from *x.y.z* to *x.(y+1).0*.)
* [ ] Does this PR change the list of algorithms available -- either adding, removing, or renaming? Does this PR otherwise change an API? (If so, PRs in fully supported downstream projects dependent on these, i.e., [oqs-provider](https://github.com/open-quantum-safe/oqs-provider) and [OQS-OpenSSH](https://github.com/open-quantum-safe/openssh) will also need to be ready for review and merge by the time this is merged.)

### CODE REVIEW

**Comment by ducnguyen-sb** [comment]
@ashman-p : I think the testvector you added is not like NIST KAT file.
E.g: 
https://github.com/open-quantum-safe/liboqs/blob/stateful-sigs/tests/KATs/sig_stfl/xmss/XMSS-SHA2_10_256.rsp

And SHA256 hash of the KAT files.
https://github.com/open-quantum-safe/liboqs/blob/stateful-sigs/tests/KATs/sig_stfl/kats.json

**Review by SWilson4 — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -56,6 +56,8 @@ The list below indicates all algorithms supported by liboqs, but not all those a
 - **Falcon**: Falcon-512, Falcon-1024
 - **SPHINCS+-SHA2**: SPHINCS+-SHA2-128f-simple, SPHINCS+-SHA2-128s-simple, SPHINCS+-SHA2-192f-simple, SPHINCS+-SHA2-192s-simple, SPHINCS+-SHA2-256f-simple, SPHINCS+-SHA2-256s-simple
 - **SPHINCS+-SHAKE**: SPHINCS+-SHAKE-128f-simple, SPHINCS+-SHAKE-128s-simple, SPHINCS+-SHAKE-192f-simple, SPHINCS+-SHAKE-192s-simple, SPHINCS+-SHAKE-256f-simple, SPHINCS+-SHAKE-256s-simple
+- **XMSS**: Variants of XMSS stateful signature scheames. e.g XMSS-SHA2_10_256, XMSSMT-SHA2_60/6, XMSSMT-SHA2_40/4_256
```
Comment (baentsch) [inline README.md:59]: Typo "schemes". Also, this location is meant to list _all_ supported algorithms with their exact names such as for users to be able to correctly request them: Are the 3 algorithms above the only XMSS ones? Are they requested with exactly these names?

File: src/sig_stfl/sig_stfl.h [hunk 2]
```diff
@@ -291,7 +294,7 @@ typedef struct OQS_SIG_STFL_SECRET_KEY {
 	/* mutual exclusion struct */
 	void *mutex;
 
-	/* file storage handle */
+	/* Application managed data related to secure storage of secret key data */
```
Comment (baentsch) [inline src/sig_stfl/sig_stfl.h:297]: How does the application have to handle this? Does it have to allocate/deallocate? Where does it pass this context in? Again, what about an example (similar to "example_kem/_sig")?

File: src/sig_stfl/sig_stfl.h [hunk 3]
```diff
@@ -338,7 +343,8 @@ typedef struct OQS_SIG_STFL_SECRET_KEY {
 	 * Callback function used to securely store key data
 	 * @param[in] sk_buf The serialized secret key data to secure store
 	 * @param[in] buf_len length of data to secure
-	 * @param[in] context aids the secure writing of data
+	 * @param[in] context application supplied data used to locate where this secret key
```
Comment (baentsch) [inline src/sig_stfl/sig_stfl.h:346]: How does the application learn how much memory is to be allocated for this?

File: src/sig_stfl/sig_stfl.h [hunk 4]
```diff
@@ -526,15 +532,35 @@ OQS_STATUS OQS_SIG_STFL_SECRET_KEY_unlock(OQS_SIG_STFL_SECRET_KEY *sk);
  *
  * @param[in] sk secret key pointer to be updated
  * @param[in] store_cb callback pointer
- * @param[in] context secret key specific data/identifier
+ * @param[in] context application data related to where how secret key data storage
```
Comment (baentsch) [inline src/sig_stfl/sig_stfl.h:535]: typo?

File: src/sig_stfl/sig_stfl.h [hunk 5]
```diff
@@ -526,15 +532,35 @@ OQS_STATUS OQS_SIG_STFL_SECRET_KEY_unlock(OQS_SIG_STFL_SECRET_KEY *sk);
...
-/* Insert stateful byte string into a secret key object. Users are responsible for deallocating buffer `sk_buf`. */
+/**
+ * OQS_SECRET_KEY_STFL_serialize_key .
+ *
+ * Insert stateful byte string into a secret key object.
+ * Users are responsible for deallocating buffer `sk_buf`.
+ *
+ * @param[in] sk secret key pointer to be serialize
+ * @param[in] sk_len size of the buffer returned
+ * @param[in] sk_buf secret key buffer returned. Caller deletes.
```
Comment (baentsch) [inline src/sig_stfl/sig_stfl.h:561]: an "in" param is returned?

**Review by baentsch — COMMENTED** [review]
I'm not sure it makes sense as part of this PR that I keep adding single comments. I did to suggest documentation improvements -- but I'd think it'd be more reasonable to do all of this as part of the upcoming main merge PR as my comments (or the changes) have nothing to do with the main topic of this PR.

File: src/sig_stfl/sig_stfl.h (hunk 3 above)
Comment (ashman-p) [reply src/sig_stfl/sig_stfl.h:346]: This called by libOQS to advance the secret key index so the the next key is used for the next sign op.

File: README.md (hunk 1 above)
Comment (ashman-p) [reply README.md:59]: Ok.

Comment (ashman-p) [reply README.md]: Updated to include the entire list.

File: src/sig_stfl/sig_stfl.h (hunk 2 above)
Comment (ashman-p) [reply src/sig_stfl/sig_stfl.h:297]: Added an example app example_sig_stfl.

File: src/sig_stfl/sig_stfl.h (hunk 3 above)
Comment (ashman-p) [reply src/sig_stfl/sig_stfl.h:346]: This function pointer points to the application's secure storage function. It is called by the oqs library. So the buffer size will be known then.

File: src/sig_stfl/sig_stfl.h (hunk 4 above)
Comment (ashman-p) [reply src/sig_stfl/sig_stfl.h]: Added the word "or" in between

File: src/sig_stfl/sig_stfl.h (hunk 5 above)
Comment (ashman-p) [reply src/sig_stfl/sig_stfl.h]: Updated.

**Review by baentsch — APPROVED** [review]
Thanks for the doc updates and the sample use, @ashman-p ! Good to merge to "stateful-sig" branch for me. Looking forward (kind of :) to review the complete PR to merge to "main".

File: src/sig_stfl/sig_stfl.c [hunk 6]
```diff
@@ -275,7 +275,11 @@ OQS_API int OQS_SIG_STFL_alg_is_enabled(const char *method_name) {
 	}
 #ifdef OQS_ENABLE_SIG_STFL_LMS
```
Comment (ducnguyen-sb) [inline src/sig_stfl/sig_stfl.c:276]: After you add individual options, you can remove `#ifdef OQS_ENABLE_SIG_STFL_LMS` like XMSS.

Comment (ashman-p) [reply src/sig_stfl/sig_stfl.c:276]: OK.

**Pull request merged**